Product Preview: Add "Preview" button to product details

ealeksandrov opened the pull request:
Closes: https://github.com/woocommerce/woocommerce-ios/issues/7944

## Description

This PR adds button on product details to show a preview. Button should appear:
- for new product flow (when it's possible to create a draft, but not for a completely blank product)
- for existing drafts (unchanged and modified).

It triggers draft saving (if needed) before opening webview with preview URL (unauthenticated at this stage).
The whole new feature is hidden behind new feature flag.

## Testing

1. Open products tab.
2. Open any existing published product.
    1. Confirm that there is no "Preview" button in navbar.
    2. Update any product property.
    3. Confirm that "Save" button appears, but there is still no "Preview" button.
3. Start the flow to create new product ("+" button in navbar on product list)
    1. Edit the title (or anything else) to have non-blank product.
    2. Confirm that "Preview" button appears in navbar.
    3. Tap it and check that draft is saved before WebView opens (ignore web content).
4. Open any existing draft product (you can continue with product from last step).
    1. Confirm that "Preview" button appears in navbar.
    4. Tap it and check that WebView immediately opens (ignore web content).
    5. Update any product property.
    6. Confirm that "Preview" button still available and tap it.
    7. Confirm that draft is saved before WebView opens.

## Screenshots

before|after
--|--
![Simulator Screen Shot](https://user-images.githubusercontent.com/3132438/198312098-b96ab486-f091-491e-9ca0-f6a9ac181023.png)|![Simulator Screen Shot](https://user-images.githubusercontent.com/3132438/198312107-a188c859-1c4f-4585-b083-4dfce05e9e9c.png)

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ealeksandrov committed:
Add completion closures to saveProduct

ealeksandrov committed:
Add products preview feature flag

ealeksandrov committed:
Add a preview button in product details

ealeksandrov committed:
Hide product preview behind feature flag

ealeksandrov committed:
Open product URL with preview parameter

ealeksandrov committed:
Update tests for actionButtons

ealeksandrov committed:
Merge branch 'trunk' into issue/7944-product-preview-flow

# Conflicts:
#	Experiments/Experiments/DefaultFeatureFlagService.swift
#	Experiments/Experiments/FeatureFlag.swift

ealeksandrov committed:
Fix save trigger condition for preview drafts

ealeksandrov committed:
Revert "Update tests for actionButtons"

223297d8d074c6bd2055c3895dbcf1b8c230d5fa

ealeksandrov commented:
Reverted unit tests on `actionButtons` because they are failing on CI due to disabled feature flag.
Will consider adding FeatureFlagService dependency injection later.

Ecarrion commented on Experiments/Experiments/DefaultFeatureFlagService.swift:
```diff
@@ -47,6 +47,8 @@ public struct DefaultFeatureFlagService: FeatureFlagService {
             return buildConfig == .localDeveloper || buildConfig == .alpha
         case .productsOnboarding:
             return buildConfig == .localDeveloper || buildConfig == .alpha
+        case .productsPreview:
+            return buildConfig == .localDeveloper || buildConfig == .alpha
```
I think we could use the same `.productsOnboarding` feature flag right? 
When the other deliverables are done, we can just take them out of the feature flag check.

ealeksandrov replied:
Right, updated in edf3702

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -199,6 +199,34 @@ final class ProductFormViewController<ViewModel: ProductFormViewModelProtocol>:
+        updatedQueryItems.append(.init(name: "preview", value: "true"))
+        permalink?.queryItems = updatedQueryItems
+        guard let url = permalink?.url else {
+            return
```
Can this fail? if it can, should we show some kind of toast?

ealeksandrov replied:
The source string, `product.permalink` is non-optional property. But all URL manipulation tools in iOS SDK return optional `URL` result.
So I consider this one is safe and never reachable.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewModel.swift:
```diff
@@ -151,6 +151,12 @@ final class ProductFormViewModel: ProductFormViewModelProtocol {
         }()
 
+        // Preview existing drafts or new products, that can be saved as a draft
+        if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productsPreview),
+           (canSaveAsDraft() || productModel.status == .draft) {
+            buttons.insert(.preview, at: 0)
+        }
```
Do you think we could move this logic to the switch above, or would things get more complicated? The switch is useful for us to not miss any case or combination. 

Also, do you think this is worth unit testing?

ealeksandrov replied:
This logic doesn't really fit in switch cases + it requires FeatureFlag check. If we add it to switch above, it'll explode into more cases and it'll be hard to understand when "preview" flow is applied.
I've updated it with comments in b6ea899, c1e812f.

Unit tests added in d69489c.

Ecarrion reviewed (commented):
The looks good, my main worry is one suggestion to make sure all cases are covered and tested!

Also I think we need to update some test cases.

----

I went through the test cases and have a couple of concerns/questions

- If you change a published product status to draft (using the products settings screen) you get the preview button, should that be the case? or should we wait until the product is actually saved as a draft before showing the preview button?

https://user-images.githubusercontent.com/562080/198395494-e68bfcbe-53a8-4704-a0a1-715bfc35b893.mov

- If you start creating a new product, you get immediately the preview button. Is that ok or should we wait for at least one change?

ealeksandrov committed:
Use productsPreview feature flag

ealeksandrov committed:
Do not preview new blank products without any changes

ealeksandrov committed:
Move comment

ealeksandrov committed:
Do not preview published products with changed status

ealeksandrov committed:
Inject FeatureFlagService dependency in ProductFormViewModel

ealeksandrov committed:
Fix ProductFormViewModelTests to use mocks with disabled feature flags

ealeksandrov committed:
Add productsOnboarding support to MockFeatureFlagService

ealeksandrov committed:
Add unit tests for preview button logic in ProductFormViewModel

ealeksandrov commented:
Thanks @Ecarrion for useful feedback and suggestions!

> Also I think we need to update some test cases.

Agree! I've added test initially, but rolled them back because of feature flag conflict.
Fixed it properly by:
- adding feature flag DI (6281f5b)
- forcing older tests to disabled flag state, so they behave in same way on CI and locally (9401ad1)
- adding `.productsOnboarding` support to MockFeatureFlagService (ac52d2d)
- adding set of tests covering preview logic under enabled feature flag (d69489c)

> If you change a published product status to draft (using the products settings screen) you get the preview button, should that be the case? or should we wait until the product is actually saved as a draft before showing the preview button?

I agree that we shouldn't allow preview here. Because it will trigger "unpublishing" of a product before user actually tap "save" to confirm a change.
Changed in 9867928.

> If you start creating a new product, you get immediately the preview button. Is that ok or should we wait for at least one change?

I don't see much issue in that, but also there is no value :)
Changed in b6ea899.

ealeksandrov committed:
Merge branch 'trunk' into issue/7944-product-preview-flow

# Conflicts:
#	Experiments/Experiments/FeatureFlag.swift
#	WooCommerce/WooCommerceTests/Mocks/MockFeatureFlagService.swift

Ecarrion reviewed (approved):
Looks great! Thanks for the changes! 🚀
